### ISSUE TITLE
[High Priority] - Element deletion bug?

### ISSUE THREAD

**Issue by kheffah** [DigitalSlideArchive/HistomicsTK#719]
When I try to delete an annotation element, or a group of annotation elements, it seems that the server request uses the `PUT` endpoint (not the `DELETE` endpoint) and when I refresh the page and reload the annotions, the elements still exist. My assumption is that the `PUT` endpoint usage is normal behavior as it tries to modify the annotation document `elements` attribute and not include the specific elements that I want to delete .. is this correct? If so, why are the elements not deleted? Keep in mind that I get a status 200 response (i.e. success) to the request. This behavior has been happening since the last update with the new Docker image.

I have recorded this behavior in developer mode in the video below:
https://drive.google.com/file/d/1l7dhrnUuoimR2jwagv0k5oSui96GFa8C/view?usp=sharing

@dgutman Do you happen to know if Styx has any weird behavior recently?

**Comment by kheffah** [comment]
@manthey I also noticed that in the "request URL", it says "annotation/{id}" without the `put` part. Could this be the bug?

**Comment by manthey** [comment]
When you delete an annotation, the annotation is deleted via the `DELETE annotation/{id}` endpoint.  When you delete elements of an annotation, the annotation is modified via the `PUT annotation/{id}` endpoint.

At least on my development box, this appears to be working correctly (elements are deleted and are not visible after a reload).

A mongo query like `db.annotationelement.find({annotationId: ObjectId('5d1b96ccd5da23f7d3ad9173')}).count()` (using the appropriate annotation ID) will give the number of elements in the annotation.  After deleting elements, this number will decrease.  Can you check if this is so?

**Comment by cooperlab** [comment]
I have encountered this before too when making our tutorial video. I could not identify a pattern but in doing a lot of takes I noticed that occasionally annotation elements that I deleted through the UI were re-appearing.

**Comment by kheffah** [comment]
@manthey I don't have direct access to the mongo database but I will ask @dgutman for help. I also noticed that in another slide the server simply gives "Internal Error" (code 500) when we try to delete an annotation. The person doing annotations also recorded this behavior but she cannot share it here because it potentially contains PHI, so I'll share the recording link to your email. 

I have a feeling this has something to do with the number of annotations in a slide and/or the sampling of annotations. She tried this with different accounts and tried to delete her own annotations versus others she has edit rights to, and none of this affects this bug. The bug seems to be strictly slide-specific .. I'm trying to find out what's in common with the slides that show this bug at the moment.

**Comment by kheffah** [comment]
@manthey I recorded the behavior in the second slide .. the one where the server request takes forever then gives Internal error.
https://drive.google.com/file/d/168NTTZ-tIFaUGI-eUZzBELf-f3H2KKXS/view

console.warn('Cannot save elements of a paged annotation');

**Comment by kheffah** [comment]
@manthey Here are the annotations from the slides with the problem:
https://drive.google.com/open?id=1aXPZpzsPN4jbWdHN4NUfRsXt4IuE_6ig
and
https://drive.google.com/open?id=1Bc0tJ7xrTJuU5llnW2GxfjnWAOaXLqY-

**Comment by kheffah** [comment]
@manthey There is something I found that may help track down the bug. 

1- For the affected slide, even though I am able to download the slide annotations using the download button in the UI, I am unable to download the annotation using an API call like:

```python
annotations = gc.get("/annotation/item/" + slide_info["_id"])
```

I get the following error:

```python
Traceback (most recent call last):

  File "<ipython-input-2-0f7b025b650a>", line 1, in <module>
    annotations = gc.get("/annotation/item/" + slide_info["_id"])

  File "/home/mtageld/anaconda3/envs/wsi/lib/python3.7/site-packages/girder_client-3.0.4.dev4-py3.7.egg/girder_client/__init__.py", line 487, in get
    return self.sendRestRequest('GET', path, parameters, jsonResp=jsonResp)

  File "/home/mtageld/anaconda3/envs/wsi/lib/python3.7/site-packages/girder_client-3.0.4.dev4-py3.7.egg/girder_client/__init__.py", line 475, in sendRestRequest
    return result.json()

  File "/home/mtageld/anaconda3/envs/wsi/lib/python3.7/site-packages/requests/models.py", line 889, in json
    self.content.decode(encoding), **kwargs

  File "/home/mtageld/anaconda3/envs/wsi/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)

  File "/home/mtageld/anaconda3/envs/wsi/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())

  File "/home/mtageld/anaconda3/envs/wsi/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None

JSONDecodeError: Expecting value
```

2- The slides with the bug contain centroid-and-size-only (CASO) annotations. The sampling shows non-editable elements as circles and does not list them when the annotation document is clicked to see a list of its elements (is this expected behavior because the elements are not loaded?). 

3- Reloading the slides with the problem occasionally shows/hides element styles in the UI. The shown/hidden styles are those from CASO annotations, but not from the "normal" polygons.

![image](https://user-images.githubusercontent.com/22067552/66707557-4a1a2d80-ed10-11e9-864a-27ae234c9ed9.png)

and after refreshing the page:

![image](https://user-images.githubusercontent.com/22067552/66707587-c4e34880-ed10-11e9-8f65-e51fe0cf8929.png)

**Comment by manthey** [comment]
@dgutman When you get a chance, can you send me the following information from this mongo instance:
```
db.annotation.find({}).count()
db.annotation.find({'_active': false}).count()
db.annotationelement.find({}).count()
```
And, what version of Mongo is currently on this instance?  This data will help replicate the problem.

I wonder if 
```
db.runCommand({compact: 'annotation'})
db.runCommand({compact: 'annotationelement'})
```
would improve performance.

**Comment by manthey** [comment]
As a response to the queries:
annotation count: 319113
annotation _active: false count: 274885
annotationelement count: 61027000

**Comment by manthey** [comment]
@dgutman I've just pushed a branch to large_image called 'old-annotations-2.x' that adds some maintenance endpoints.  `GET annotation/old` takes an age in days and a number of history states, and returns how many annotation history states are older than that.  The companion `DELETE` endpoint will remove old annotation history and compact the database.  `DELETE` won't let you delete anything active or younger than 7 days.  You should be able to cherry pick that branch to whatever version you are running to at least see what the `GET` endpoint reports.

**Comment by manthey** [comment]
This underlying problem was fixed in #761, so I'll close this issue.  https://github.com/girder/large_image/pull/401 may help speed up the database.
